pydantic/_internal/_dataclasses.py

```diff
@@ -92,10 +95,22 @@ def prepare_dataclass(
 
     core_config = generate_config(config, cls)
     cls.__pydantic_fields__ = fields
-    cls.__pydantic_validator__ = SchemaValidator(schema, core_config)
+    cls.__pydantic_validator__ = validator = SchemaValidator(schema, core_config)
     # this works because cls has been transformed into a dataclass by the time "cls" is called
     cls.__pydantic_serializer__ = SchemaSerializer(schema, core_config)
 
+    extra = config.get('extra', Extra.forbid)
+    if extra != Extra.forbid:
+        raise ValueError(f'extra=Extra.{extra.value} is not allowed for dataclasses. Only Extra.forbid is allowed.')
```

Comment:
I'm curious the reasoning behind removing support for this.

I think it's fairly common in FastAPI world to want to allow extra fields to be passed into your API (e.g. so you can remove ignored fields from your model without it causing errors), and there have been many issues over the years of people wanting to use dataclasses instead of BaseModel in their fastapi endpoints. 

And today, you can allow extra for dataclasses in FastAPI:
```
import pydantic.dataclasses
from dataclasses import dataclass

from fastapi import FastAPI
from fastapi.testclient import TestClient

app = FastAPI()


@pydantic.dataclasses.dataclass
class MyBody:
    x: int


@app.post("/items/", include_in_schema=False)
async def read_items(body: MyBody):
    return {"x": body.x}


client = TestClient(app)

response = client.post("/items/", json={'x': 1, "y": 2})
assert response.status_code == 200, response.text
assert response.json() == {"x": 1}, response.json()
```

I would be amenable to the idea of dropping support for this if it causes serious issues (or if the plan is to do something different for FastAPI validation, somehow). I think if we want to make this change though and won't support extra fields for FastAPI body arguments annotated with dataclasses, we should probably run that by Sebastian for feedback.

Reply:
I realized I wasn't clear in some of my wording above: I think it is important to support `extra='ignore'`, I am fine dropping support for `extra='allow'` for dataclasses.

Reply:
Made the default `ignore` and prohibited just `allow`. Thanks folks.